kartothek/core/dataset.py

```diff
@@ -907,6 +907,48 @@ def from_dataset(dataset):
         ds_builder.partitions = dataset.partitions
         return ds_builder
 
+    def modify_uuid(self, target_uuid: str):
+        """
+        Modify the dataset uuid and depending metadata:
+        - paths to partitioning files
+        - path to index files
+
+        Parameters
+        ----------
+        target_uuid: str
+            Modified dataset UUID.
+
+        Returns
+        -------
+        DatasetMetadataBuilder
+            modified builder object
+        """
+
+        # modify file names in partition metadata
+        modified_partitions = {}
+        for p_key, p in self.partitions.items():
+            pdict = p.to_dict()
+            for table_key, table_file in pdict["files"].items():
+                if table_file.startswith(f"{self.uuid}/"):
+                    pdict["files"][table_key] = table_file.replace(
+                        self.uuid, target_uuid, 1
+                    )
+            modified_partitions[p_key] = Partition.from_dict(p_key, pdict)
+
+        self.partitions = modified_partitions
+
+        for i_key, i in self.indices.items():
+            if (hasattr(i, "index_storage_key")) and (
+                getattr(i, "index_storage_key") is not None
+            ):
+                setattr(
+                    i,
+                    "index_storage_key",
+                    getattr(i, "index_storage_key").replace(self.uuid, target_uuid, 1),
+                )
```

Comment:
Instead of checking for the attribute, I would suggest to check for the type. imo, there is no need for duck-typing here